src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -38,37 +51,232 @@ public class PackageMetadataResourceV3 : PackageMetadataResource
+            var registrationUri = _regResource.GetUri(packageId);
+
+            var registrationIndexResult = await LoadRegistrationIndexAsync(
+                _client,
+                registrationUri,
+                packageId,
+                sourceCacheContext,
+                httpSourceResult => DeserializeStreamDataAsync<RegistrationIndex>(httpSourceResult.Stream, token),
+                log,
+                token);
+
+            var registrationIndex = registrationIndexResult.Index;
+            var httpSourceCacheContext = registrationIndexResult.CacheContext;
+
+            if (registrationIndex == null)
+            {
+                // The server returned a 404, the package does not exist
+                return Enumerable.Empty<PackageSearchMetadataRegistration>();
+            }
+
+            var results = new List<PackageSearchMetadataRegistration>();
+
+            foreach (var registrationPage in registrationIndex.Items)
+            {
+                if (registrationPage == null)
+                {
+                    throw new InvalidDataException(registrationUri.AbsoluteUri);
+                }
+
+                var lower = NuGetVersion.Parse(registrationPage.Lower);
+                var upper = NuGetVersion.Parse(registrationPage.Upper);
+
+                if (range.DoesRangeSatisfy(lower, upper))
+                {
+                    if (registrationPage.Items == null)
+                    {
+                        var rangeUri = registrationPage.Url;
+                        var leafRegistrationPage = await GetRegistratioIndexPageAsync(_client, rangeUri, packageId, lower, upper, httpSourceCacheContext, log, token);
+
+                        if (registrationPage == null)
+                        {
+                            throw new InvalidDataException(registrationUri.AbsoluteUri);
+                        }
+
+                        ProcessRegistrationPage(leafRegistrationPage, results, range, includePrerelease,includeUnlisted, metadataCache);
+                    }
+                    else
+                    {
+                        ProcessRegistrationPage(registrationPage, results, range, includePrerelease, includeUnlisted, metadataCache);
+                    }
+                }
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Deserialize stream from RegistrationIndex/RegistrationPage and return list of RegistrationPages or RegistrationPage.
+        /// </summary>
+        /// <typeparam name="T">Generic type</typeparam>
+        /// <param name="stream">Stream data to read.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns></returns>
+        private async Task<T> DeserializeStreamDataAsync<T>(Stream stream, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            var jsonSerializer = JsonSerializer.Create(JsonExtensions.ObjectSerializationSettings);
```

Comment:
I'd suggest making `JsonExtneions.JsonObjectSerializer` internal, then reuse that instance to avoid allocating, and therefore garbage collecting, a new `JsonSerializer` instance each time.

Reply:
Changed.